### PR TITLE
Allow extensions to modify also dynamic parts of collector config

### LINKED ISSUE #2001: Custom metrics pipelines are not visible by fluxninja extension
**Describe the bug**
User-defined metrics (`metrics/user-defined-*`) is not visible by fluxninja plugin and fluxninja exporters are not added to these metrics (with the exception of `metrics/user-defined-kubeletstats` which works)

**Steps to reproduce**
```
nice tilt up -- --race --manual --scenario scenarios/rabbitmq-queue-buildup/
```
(needs some tweaks to force-enable cloud plugin without api key)

**What did you expect to see?**
`addFNToPipeline` called on `metrics/user-defined-rabbitmq`.

**What did you see instead?**
Not called. Also, the hook seems to be called only once, so it doesn't have any chance to be called on infra meters added dynamically by policies.

**What version did you use?**
44e6df4ac1715bbf4d98589f592d761722f3266c

**Additional context**
Might be caused by https://github.com/fluxninja/aperture/pull/1890

### PR DESCRIPTION
**Pull request by krdln**
### Description of change

Allow extensions to modify also dynamic parts of collector config.

In previous approach, extensions were able to provide their own part of
collector config only based on base config, with no ability to depend on
dynamic parts of config. This was preventing eg. fluxninja extesion to
add fluxninja exporters to custom metric pipelines (infra meters).

This commit reverts the idea of multiple config providers. Extension
now provide their config by "hooks" instead. Hooks are executed on every
config udpate, so now extensions can modify also the dynamically
injected parts of config.

Drive-by:
* Actually wait for collector's shutdown in its stop hook.
* Rename structs in otelcollector/config package to avoid stutter.
* Add some logging to fluxninja extension.

Resolves #2001

##### Checklist

- [x] Tested in playground or other setup
- [x] Tested in playground or other setup (with cloud)
- [x] Screenshot (Grafana) from playground added to PR for 15+ minute run
- [x] Tests and/or benchmarks are included
- [x] Fix tests
- [x] Remove temporary hacks / logs.

### Summary by OpenAI

**New Feature:**
- Extensions can now modify dynamic parts of the collector config through hooks
- Introduced `Provider` struct for updating config and registering hooks

**Refactor:**
- Renamed structs and package names for better clarity
- Ensured collector starts after all extensions have started

**Test:**
- Added test files for the `otelconfig` package and new features

> 🎉 A new dawn breaks, with hooks in place,
> Dynamic configs, extensions embrace. 🌟
> Structs renamed, clarity achieved,
> Collector waits, as extensions breathe. 🚀

### CODE REVIEW

File: extensions/fluxninja/otel/provide.go [hunk 1]
```diff
@@ -38,74 +37,67 @@ const (
...
+		// We must add the config-modifying hook in this start hook, as we need
+		// heartbeats.ControllerInfo, that's populated only in heartbeats start.
+		configProvider.AddMutatingHook(func(config *otelconfig.Config) {
+			addFluxNinjaExporter(config, extensionConfig, grpcClientConfig, httpClientConfig)
 
-	lifecycle.Append(fx.Hook{
-		OnStart: func(context.Context) error {
-			controllerID := heartbeats.ControllerInfo.Id
+			controllerID := "XXX no heartbeats?"
+			// controllerID := heartbeats.ControllerInfo.Id
```
Comment (krdln) [inline extensions/fluxninja/otel/provide.go:73]: Remember to revert this :upside_down_face:

File: cmd/aperture-agent/agent/provider.go [hunk 2]
```diff
@@ -21,40 +19,39 @@ func ProvidePeersPrefix(agentInfo *agentinfo.AgentInfo) (peers.PeerDiscoveryPref
...
-	cfg.AddProcessor(otelconsts.ProcessorAgentResourceLabels, map[string]interface{}{
-		"log_statements":    transformStatements,
-		"metric_statements": transformStatements,
+		}
+		cfg.AddProcessor(otelconsts.ProcessorAgentResourceLabels, map[string]interface{}{
+			"log_statements":    transformStatements,
+			"metric_statements": transformStatements,
+		})
 	})
 }
```
[the remark this answers is not in the document]
Comment (krdln) [reply cmd/aperture-agent/agent/provider.go:57]: This function is infallible.

File: pkg/otelcollector/config/config-provider.go [hunk 3]
```diff
@@ -1,79 +1,128 @@
...
-	u.lock.Lock()
-	defer u.lock.Unlock()
-	return u.config
+// MustGetConfig returns a snapshot of the current config.
+func (p *Provider) MustGetConfig() *Config {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	// Copying to avoid concurrent modification by hooks.
+	return copystructure.Must(copystructure.Copy(p.config)).(*Config)
 }
```
[the remark this answers is not in the document]
Comment (krdln) [reply pkg/otelcollector/config/config-provider.go]: Both methods call setConfig with the lock held, so there's no issue. In your suggested code, `p.config` and `p,hooks` are accessed unprotected though.

File: pkg/otelcollector/config/provider.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	p.setConfig(p.config)
+}
+
+// MustGetConfig returns a snapshot of the current config.
+func (p *Provider) MustGetConfig() *Config {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	// Copying to avoid concurrent modification by hooks.
+	return copystructure.Must(copystructure.Copy(p.config)).(*Config)
+}
```
[the remark this answers is not in the document]
Comment (krdln) [reply pkg/otelcollector/config/provider.go:128]: All hooks are executed serially, so there's no risk of concurrent modification. Also, they're intentionally allowed to modify config.

File: pkg/otelcollector/otelcollector.go [hunk 5]
```diff
@@ -113,19 +101,40 @@ func setup(in ConstructorIn) error {
...
+				cancelRun()
+			}()
+
 			otelService.Shutdown()
+			// Shutdown only starts the shutdown process, we need to wait for
+			// actual shutdown to finish.
+			wg.Wait()
 			return nil
 		},
 	})
```
[the remark this answers is not in the document]
Comment (krdln) [reply pkg/otelcollector/otelcollector.go:140]: No, in fx lifecycle the OnStop hook will only be called after OnStart and only after OnStart succeeds

File: cmd/aperture-agent/agent/otel-config.go [hunk 6]
```diff
@@ -56,19 +56,22 @@ func provideAgent(
...
 	allInfraMeters := map[string]map[string]*policylangv1.InfraMeter{}
 	handleInfraMeterUpdate := func(event notifiers.Event, unmarshaller config.Unmarshaller) {
+		var err error //nolint:govet
 		log.Info().Str("event", event.String()).Msg("infra meter update")
 		tc := &policylangv1.TelemetryCollector{}
-		if err := unmarshaller.UnmarshalKey("", tc); err != nil {
+		if err = unmarshaller.UnmarshalKey("", tc); err != nil {
 			log.Error().Err(err).Msg("unmarshalling telemetry collector")
 			return
 		}
```
[the remark this answers is not in the document]
Comment (krdln) [reply cmd/aperture-agent/agent/otel-config.go:77]: No, because govet complains about shadowing

**Comment by krdln** [comment]
image

File: test/aperture_suite_test.go [hunk 7]
```diff
@@ -275,5 +268,5 @@ func provideOTelConfig() *otelconfig.OTelConfigProvider {
 			Exporters: []string{otelconsts.ExporterLogging},
 		})
 	}
-	return otelconfig.NewOTelConfigProvider("service", cfg)
+	return otelconfig.NewProvider("service", cfg)
 }
```
[the remark this answers is not in the document]
Comment (krdln) [reply test/aperture_suite_test.go:272]: There are no multiple goroutines here

File: extensions/fluxninja/otel/provide.go [hunk 8]
```diff
@@ -169,8 +161,8 @@ func addMetricsControllerSlowPipeline(baseConfig, config *otelconfig.OTelConfig)
 	})
 }
 
-func addFluxNinjaPrometheusReceiver(baseConfig, config *otelconfig.OTelConfig) {
-	rawReceiverConfig, _ := baseConfig.Receivers[otelconsts.ReceiverPrometheus].(map[string]any)
+func addFluxNinjaPrometheusReceiver(config *otelconfig.Config) {
+	rawReceiverConfig, _ := config.Receivers[otelconsts.ReceiverPrometheus].(map[string]any)
 	duplicatedReceiverConfig, err := duplicateMap(rawReceiverConfig)
 	if err != nil {
 		// It should not happen, unless the original config is messed up.
```
[the remark this answers is not in the document]
Comment (krdln) [reply extensions/fluxninja/otel/provide.go:168]: it is handled via log.Fatal

**Review by kwapik — APPROVED** [review]
Pretty cool idea with hooks :+1:

File: cmd/aperture-controller/controller/otel-config.go [hunk 9]
```diff
@@ -17,27 +17,26 @@ func provideController(
...
+	return otelconfig.NewProvider("service", otelCfg), nil
 }
 
 // addMetricsPipeline adds metrics to pipeline for controller OTel collector.
 func addMetricsPipeline(
-	config *otelconfig.OTelConfig,
+	config *otelconfig.Config,
 	controllerConfig *controllerconfig.ControllerOTelConfig,
 	tlsConfig *tls.Config,
 	lis *listener.Listener,
```
[the remark this answers is not in the document]
Comment (krdln) [reply cmd/aperture-controller/controller/otel-config.go:42]: it does modify.

File: cmd/aperture-controller/controller/otel-config.go [hunk 10]
```diff
@@ -53,7 +52,7 @@ func addMetricsPipeline(
 }
 
 func addPrometheusReceiver(
-	config *otelconfig.OTelConfig,
+	config *otelconfig.Config,
 	controllerConfig *controllerconfig.ControllerOTelConfig,
 	tlsConfig *tls.Config,
 	lis *listener.Listener,
```
[the remark this answers is not in the document]
Comment (krdln) [reply cmd/aperture-controller/controller/otel-config.go:58]: it does modify the config.

**Pull request merged**